File: golang/uPIMulator/src/simulator/dpu/logic/operand_collector.go
package logic

import (
	"errors"
	"uPIMulator/src/abi/encoding"
	"uPIMulator/src/abi/word"
	"uPIMulator/src/simulator/dpu/sram"
)

type OperandCollector struct {
	wram *sram.Wram
}

func (this *OperandCollector) Init() {
	this.wram = nil
}

func (this *OperandCollector) Fini() {
}

func (this *OperandCollector) ConnectWram(wram *sram.Wram) {
	if this.wram != nil {
		err := errors.New("wram is already set")
		panic(err)
	}

	this.wram = wram
}

func (this *OperandCollector) Lbs(address int64) int64 {
	byte_stream := this.wram.Read(address, 1)

	value := int64(byte_stream.Get(0))

	word_ := new(word.Word)
	word_.Init(8)
	word_.SetValue(value)

	return word_.Value(word.SIGNED)
}

func (this *OperandCollector) Lbu(address int64) int64 {
	byte_stream := this.wram.Read(address, 1)

	value := int64(byte_stream.Get(0))

	word_ := new(word.Word)
	word_.Init(8)
	word_.SetValue(value)

	return word_.Value(word.UNSIGNED)
}

func (this *OperandCollector) Lhs(address int64) int64 {
	word_ := new(word.Word)
	word_.Init(16)
	word_.SetBitSlice(0, 8, this.Lbs(address))
	word_.SetBitSlice(8, 16, this.Lbs(address+1))
	return word_.Value(word.SIGNED)
}

func (this *OperandCollector) Lhu(address int64) int64 {
	word_ := new(word.Word)
	word_.Init(16)
	word_.SetBitSlice(0, 8, this.Lbu(address))
	word_.SetBitSlice(8, 16, this.Lbu(address+1))
	return word_.Value(word.UNSIGNED)
}

func (this *OperandCollector) Lw(address int64) int64 {
	word_ := new(word.Word)
	word_.Init(32)
	word_.SetBitSlice(0, 8, this.Lbu(address))
	word_.SetBitSlice(8, 16, this.Lbu(address+1))
	word_.SetBitSlice(16, 24, this.Lbu(address+2))
	word_.SetBitSlice(24, 32, this.Lbu(address+3))
	return word_.Value(word.UNSIGNED)
}

func (this *OperandCollector) Ld(address int64) (int64, int64) {
	return this.Lw(address + 4), this.Lw(address)
}

func (this *OperandCollector) Sb(address int64, value int64) {
	word_ := new(word.Word)
	word_.Init(8)
	word_.SetValue(value)

	byte_stream := new(encoding.ByteStream)
	byte_stream.Init()
	byte_stream.Append(uint8(word_.Value(word.UNSIGNED)))

	this.wram.Write(address, 1, byte_stream)
}

func (this *OperandCollector) Sh(address int64, value int64) {
	word_ := new(word.Word)
	word_.Init(16)
	word_.SetValue(value)

	this.Sb(address, word_.BitSlice(word.UNSIGNED, 0, 8))
	this.Sb(address+1, word_.BitSlice(word.UNSIGNED, 8, 16))
}

func (this *OperandCollector) Sw(address int64, value int64) {
	word_ := new(word.Word)
	word_.Init(32)
	word_.SetValue(value)

	this.Sb(address, word_.BitSlice(word.UNSIGNED, 0, 8))
	this.Sb(address+1, word_.BitSlice(word.UNSIGNED, 8, 16))
	this.Sb(address+2, word_.BitSlice(word.UNSIGNED, 16, 24))
	this.Sb(address+3, word_.BitSlice(word.UNSIGNED, 24, 32))
}

func (this *OperandCollector) Sd(address int64, even int64, odd int64) {
	this.Sw(address+4, even)
	this.Sw(address, odd)
}

File: python_cpp/uPIMulator_backend/src/simulator/dpu/operand_collector.cc
#include "simulator/dpu/operand_collector.h"

#include <tuple>

namespace upmem_sim::simulator::dpu {

void OperandCollector::connect_wram(sram::WRAM *wram) {
  assert(wram != nullptr);
  assert(wram_ == nullptr);

  wram_ = wram;
}

int64_t OperandCollector::lbs(Address address) {
  Address data_word_size = abi::word::DataWord().size();
  Address base_address = (address / data_word_size) * data_word_size;
  Address offset = address % data_word_size;

  auto data_word = new abi::word::DataWord();
  data_word->set_value(wram_->read(base_address));

  int64_t result =
      data_word->bit_slice(abi::word::SIGNED, static_cast<int>(8 * offset),
                           static_cast<int>(8 * (offset + 1)));

  delete data_word;

  return result;
}

int64_t OperandCollector::lbu(Address address) {
  Address data_word_size = abi::word::DataWord().size();
  Address base_address = (address / data_word_size) * data_word_size;
  Address offset = address % data_word_size;

  auto data_word = new abi::word::DataWord();
  data_word->set_value(wram_->read(base_address));

  int64_t result =
      data_word->bit_slice(abi::word::UNSIGNED, static_cast<int>(8 * offset),
                           static_cast<int>(8 * (offset + 1)));

  delete data_word;

  return result;
}

int64_t OperandCollector::lhs(Address address) {
  auto data_word = new abi::word::DataWord();

  data_word->set_bit_slice(0, 8, lbs(address));
  data_word->set_bit_slice(8, 16, lbs(address + 1));

  int64_t result = data_word->bit_slice(abi::word::SIGNED, 0, 16);

  delete data_word;

  return result;
}

int64_t OperandCollector::lhu(Address address) {
  auto data_word = new abi::word::DataWord();

  data_word->set_bit_slice(0, 8, lbu(address));
  data_word->set_bit_slice(8, 16, lbu(address + 1));

  int64_t result = data_word->bit_slice(abi::word::UNSIGNED, 0, 16);

  delete data_word;

  return result;
}

int64_t OperandCollector::lw(Address address) {
  auto data_word = new abi::word::DataWord();

  data_word->set_bit_slice(0, 8, lbu(address));
  data_word->set_bit_slice(8, 16, lbu(address + 1));
  data_word->set_bit_slice(16, 24, lbu(address + 2));
  data_word->set_bit_slice(24, 32, lbu(address + 3));

  int64_t result = data_word->value(abi::word::UNSIGNED);

  delete data_word;

  return result;
}

std::tuple<int64_t, int64_t> OperandCollector::ld(Address address) {
  return {lw(address + abi::word::DataWord().size()), lw(address)};
}

void OperandCollector::sb(Address address, int64_t value) {
  Address data_word_size = abi::word::DataWord().size();
  Address base_address = (address / data_word_size) * data_word_size;
  Address offset = address % data_word_size;

  auto data_word = new abi::word::DataWord();
  data_word->set_value(wram_->read(base_address));
  data_word->set_bit_slice(static_cast<int>(8 * offset),
                           static_cast<int>(8 * (offset + 1)), value);

  wram_->write(base_address, data_word->value(abi::word::UNSIGNED));

  delete data_word;
}

void OperandCollector::sh(Address address, int64_t value) {
  auto data_word = new abi::word::DataWord();
  data_word->set_value(value);

  sb(address, data_word->bit_slice(abi::word::UNSIGNED, 0, 8));
  sb(address + 1, data_word->bit_slice(abi::word::UNSIGNED, 8, 16));

  delete data_word;
}

void OperandCollector::sw(Address address, int64_t value) {
  auto data_word = new abi::word::DataWord();
  data_word->set_value(value);

  sb(address, data_word->bit_slice(abi::word::UNSIGNED, 0, 8));
  sb(address + 1, data_word->bit_slice(abi::word::UNSIGNED, 8, 16));
  sb(address + 2, data_word->bit_slice(abi::word::UNSIGNED, 16, 24));
  sb(address + 3, data_word->bit_slice(abi::word::UNSIGNED, 24, 32));

  delete data_word;
}

void OperandCollector::sd(Address address, int64_t even, int64_t odd) {
  sw(address + abi::word::DataWord().size(), even);
  sw(address, odd);
}

}  // namespace upmem_sim::simulator::dpu
